Checks the ETag of the files being requested to prevent downloading a…

Tro95 opened the pull request:
Checks the ETag of the files being requested to prevent downloading and replacing files with no updates. https://github.com/aws-cloudformation/cfn-python-lint/issues/1249

*Issue #, if available:* 1249

*Description of changes:* Added caching functionality to the `get_url_content()` helper function, that stores and compares ETag values in the response headers.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Tro95 commented on src/cfnlint/helpers.py:
```diff
@@ -165,10 +166,52 @@
     'AWS::Redshift::Cluster'
 ]
 
+CACHING_DIR = os.path.join(os.path.dirname(__file__), 'cache/')
```
This can easily be changed to a different directory or moved into a config option if desired. I'm also unsure of the consequences of using `__file__` rather than `cfnlint.__file__` as is used in some places, but there doesn't seem to be a consistent approach being used, and there were circular dependency issues with the latter option.

kddejong replied:
Wondering if it should be under the data folder?  Would we want to commit that file as we do the `update-specs` so the most recent version is included in the package?

Tro95 replied:
I'm happy for it to live under the data folder; the primary reason for using a separate cache directory was just to signify the temporary, non-essential nature of the files. It definitely could be commited and used like a lock file, and would save an initial fetch of every spec file when it's first run.

Tro95 commented on test/unit/module/helpers/test_get_url_content.py:
```diff
@@ -11,6 +11,9 @@
 from mock import patch, MagicMock
 import cfnlint.helpers
+import random
+import string
```
These can both be removed if people were fine with hard-coding the ETag value in the tests. I just used them to generate a random string.

Tro95 commented:
With this pull request I'm seeing a reduction from ~33 seconds to download all the specifications to ~20 seconds when none of them need redownloading. Not exactly the kind of increase I was hoping for, but still an improvement, and I'm sure someone can make it even better.

PatMyron commented:
> With this pull request I'm seeing a reduction from ~33 seconds to download all the specifications to ~20 seconds when none of them need redownloading. Not exactly the kind of increase I was hoping for, but still an improvement, and I'm sure someone can make it even better.

Especially as the number of regions continues to grow, I think updating the regions in parallel will be the main key to keeping `cfn-lint --update-specs` completion times within a couple seconds

https://github.com/aws-cloudformation/cfn-python-lint/blob/52864add9eac85a42cc6866cdf05c4d3b4dcd149/src/cfnlint/maintenance.py#L50

PatMyron commented:
> With this pull request I'm seeing a reduction from ~33 seconds to download all the specifications to ~20 seconds when none of them need redownloading. Not exactly the kind of increase I was hoping for, but still an improvement, and I'm sure someone can make it even better.

Especially as the number of regions continues to grow, I think updating the regions in parallel will be the main key to keeping `cfn-lint --update-specs` completion times within a couple seconds

https://github.com/aws-cloudformation/cfn-python-lint/blob/52864add9eac85a42cc6866cdf05c4d3b4dcd149/src/cfnlint/maintenance.py#L50

kddejong commented:
I think parallelization is a great idea here.  Another option would be to combine the region and update specs together.  Allowing for a user to filter down to the specs they care about.  By default we would have to keep the all which may make this a little trickier to implement.

Tro95 commented:
I'm happy to give paralleisation a go, is there any particular parellelisation module you'd prefer to use? The constant I/O of the downloads metadata file concerns me, both from a mutex point-of-view, and the inefficiency of reading and writing a whole file for a single line change. What are your thoughts on having an individual metadata file per a canonical url, i.e. `<sha256 of url>.meta.json` holding data specifically related to that url, rather than a single, centralised file?

Tro95 commented:
I've added parallelisation and changed some basic things like the directory the caching files were being pulled into.

Updating all files speed:
```
$ time cfn-lint --update-specs

real	0m11.139s
user	0m22.762s
sys	0m2.131s
```
Subsequent update with no specs to update (as per cache check):
```
$ time cfn-lint --update-specs

real	0m5.726s
user	0m11.829s
sys	0m1.781s
```

Tro95 commented on src/cfnlint/maintenance.py:
```diff
@@ -21,19 +22,37 @@
 
-        with open(filename, 'w') as f:
-            json.dump(spec, f, indent=2, sort_keys=True, separators=(',', ': '))
+    multiprocessing_logger = multiprocessing.log_to_stderr()
```
One thing to note is I couldn't get logging to work inside the parallelised processes. Everything inside the parallel process is only debug logging, however it's definitely something that could be improved upon.

miparnisari replied:
Yesterday night  [I tried out a similar implementation to yours, but using multiple threads](https://github.com/aws-cloudformation/cfn-python-lint/pull/1410), and I think logging works there.

miparnisari replied:
I saw logs so I'm resolving this conversation.

PatMyron commented:
Is `cfn-lint --update-specs` still generating [`src/cfnlint/data/CloudSpecs/`](https://github.com/aws-cloudformation/cfn-python-lint/tree/master/src/cfnlint/data/CloudSpecs) changes [like the ones here](https://github.com/aws-cloudformation/cfn-python-lint/pull/1355)? I wasn't able to generate any using this branch:

```shell
$ pip3 install -e . --user
$ cfn-lint --update-specs
$ git status
On branch cfn-issue-1249
Your branch is up to date with 'origin/cfn-issue-1249'.
nothing to commit, working tree clean
```

Tro95 commented:
It should be making changes (assuming there are any to make), so I'm unsure why it's not generating any for you.
```
$ git show
commit 706369e0e0aacbe90c150c1a7b761d880ce601c4 (HEAD -> cfn-issue-1249, dev/cfn-issue-1249)
Merge: ab46e27a be3867c6
Author: Kevin DeJong <kddejong@amazon.com>
Date:   Sat Mar 14 07:25:39 2020 -0500

    Merge branch 'master' into cfn-issue-1249

$ git status
On branch cfn-issue-1249
Your branch is up to date with 'dev/cfn-issue-1249'.

nothing to commit, working tree clean
$ pip3 install -e . --user
$ cfn-lint --update-specs
$ git status
On branch cfn-issue-1249
Your branch is up to date with 'dev/cfn-issue-1249'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   src/cfnlint/data/CloudSpecs/ap-east-1.json
	modified:   src/cfnlint/data/CloudSpecs/ap-northeast-1.json
	modified:   src/cfnlint/data/CloudSpecs/ap-northeast-2.json
	modified:   src/cfnlint/data/CloudSpecs/ap-northeast-3.json
	modified:   src/cfnlint/data/CloudSpecs/ap-south-1.json
	modified:   src/cfnlint/data/CloudSpecs/ap-southeast-1.json
	modified:   src/cfnlint/data/CloudSpecs/ap-southeast-2.json
	modified:   src/cfnlint/data/CloudSpecs/ca-central-1.json
	modified:   src/cfnlint/data/CloudSpecs/cn-north-1.json
	modified:   src/cfnlint/data/CloudSpecs/cn-northwest-1.json
	modified:   src/cfnlint/data/CloudSpecs/eu-central-1.json
	modified:   src/cfnlint/data/CloudSpecs/eu-north-1.json
	modified:   src/cfnlint/data/CloudSpecs/eu-west-1.json
	modified:   src/cfnlint/data/CloudSpecs/eu-west-2.json
	modified:   src/cfnlint/data/CloudSpecs/eu-west-3.json
	modified:   src/cfnlint/data/CloudSpecs/me-south-1.json
	modified:   src/cfnlint/data/CloudSpecs/sa-east-1.json
	modified:   src/cfnlint/data/CloudSpecs/us-east-1.json
	modified:   src/cfnlint/data/CloudSpecs/us-east-2.json
	modified:   src/cfnlint/data/CloudSpecs/us-gov-east-1.json
	modified:   src/cfnlint/data/CloudSpecs/us-gov-west-1.json
	modified:   src/cfnlint/data/CloudSpecs/us-west-1.json
	modified:   src/cfnlint/data/CloudSpecs/us-west-2.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/cfnlint/data/DownloadsMetadata/

no changes added to commit (use "git add" and/or "git commit -a")
```

PatMyron commented:
@Tro95 sorry, I forgot I was testing the [homebrew](https://github.com/aws-cloudformation/cfn-python-lint/issues/1434) version recently, was able to generate changes here after uninstalling that

kddejong commented:
@PatMyron this looking good to merge?

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
+        # If we have an ETag value stored and it matches the returned one,
+        # then we already have a copy of the most recent version of the
+        # resource, so don't bother fetching it again
+        if etag and res.info().get('ETag'):
+            if etag == res.info().get('ETag'):
```
You can combine the two `if`s.

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
+        # If we have an ETag value stored and it matches the returned one,
+        # then we already have a copy of the most recent version of the
+        # resource, so don't bother fetching it again
+        if etag and res.info().get('ETag'):
```
Suggestion: rename `etag` to `cached_etag`

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
+
+        # Write the etag back to our cache
+        if res.info().get('ETag'):
+            metadata['etag'] = res.info().get('ETag')
```
Suggestion: rename `metadata` to `cache`

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
+        if res.info().get('ETag'):
+            metadata['etag'] = res.info().get('ETag')
+            metadata['url'] = url # To make it obvious which url the Tag relates to
+            save_download_metadata(metadata, metadata_filename)
```
Suggestion: rename function to `update_cache`

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
-def get_url_content(url):
+
+def get_url_content(url, caching=False):
     """Get the contents of a spec file"""
```
To be more specific, the function returns `None` if the spec file is in the cache.

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
+
+    if caching:
+        # Load in the cache
+        metadata = get_download_metadata(metadata_filename)
```
Suggestion, rename function to `load_cache`

Tro95 replied:
I prefer the term metadata because what's being stored isn't the cached file itself, but metadata about the file that gets used to check if we have the latest version. There may be future use cases where we want to store additional metadata about the files that isn't related to caching.

miparnisari replied:
Yeah, `load_metadata` sounds better

miparnisari commented on src/cfnlint/helpers.py:
```diff
@@ -166,10 +167,63 @@
 
-def get_url_content(url):
+
+def get_url_content(url, caching=False):
```
This function is doing too many things... it's loading up the cache, checking to see if an update needs to happen, and if it does, it is downloading the file, and then updating the cache. Can we split it up into smaller chunks?

miparnisari commented on src/cfnlint/maintenance.py:
```diff
@@ -21,19 +22,43 @@
+
+    # Check to see if no content was returned, indicating we have the latest version already
+    if not spec_content:
+        multiprocessing_logger.debug('We already have the most recent version of %s stored in %s', url, filename)
```
Is this message necessary? `get_url_content` is already logging this

miparnisari commented on src/cfnlint/maintenance.py:
```diff
@@ -21,19 +22,43 @@
+
+    spec_content = get_url_content(url, caching=True)
+
+    # Check to see if no content was returned, indicating we have the latest version already
```
This comment hints that there could be a function `should_perform_update` that returns true if the cache is old, false otherwise

Tro95 replied:
Good call, I separated the update logic into a separate method `url_has_newer_version`.

Tro95 commented on src/cfnlint/data/DownloadsMetadata/1c9ead4af49b3a8f39632f5a30578ead5310da0b5a68ae4cf93b4be6a9a05278.meta.json:
```diff
@@ -0,0 +1 @@
+{"etag": "\"f71b5ea17b03894692ac60518e30db24\"", "url": "https://s3.us-gov-west-1.amazonaws.com/cfn-resource-specifications-us-gov-west-1-prod/latest/gzip/CloudFormationResourceSpecification.json"}
```
It makes sense to commit them, otherwise someone with the latest specs running `--update-specs` would be forced to redownload them one time to generate these files.

miparnisari replied:
> someone with the latest specs running --update-specs would be forced to redownload them one time to generate these files

Having people download the files one more time is a small price to pay IMO. I prefer to keep the codebase clean. They'll reap the benefits later :)

Tro95 replied:
I'll remove them from this PR, but I think we should consider adding them in future. If someone is using cfn-pythn-lint in a CI/CD flow where they pull the latest version each build, running `--update-specs` will force a download every time even though they might already have the latest.

PatMyron commented:
@kddejong @miparnisari if the [resource specifications](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html) haven't changed, some of the other scripts we run won't be applied without modifying [`src/cfnlint/data/DownloadsMetadata`](https://github.com/aws-cloudformation/cfn-python-lint/tree/master/src/cfnlint/data/DownloadsMetadata):
```shell
pip3 install -e .
scripts/update_specs_from_pricing.py # requires Boto3 and Credentials
cfn-lint --update-specs # won't apply changes if metadata matches
```

Tro95 commented:
> @kddejong @miparnisari if the [resource specifications](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html) haven't changed, some of the [other data sources](https://github.com/aws-cloudformation/cfn-python-lint/pull/1480) won't be applied without modifying [`src/cfnlint/data/DownloadsMetadata`](https://github.com/aws-cloudformation/cfn-python-lint/tree/master/src/cfnlint/data/DownloadsMetadata):
> 
> ```shell
> pip3 install -e .
> scripts/update_specs_from_pricing.py # requires Boto3 and Credentials
> cfn-lint --update-specs # won't apply changes if metadata matches
> ```

@PatMyron Is the problem the fact that `--update-specs` does both the downloading of updated specs, and the merging of the specs with the patches produced by the scripts? If so, the fix should be relatively straight-forward, by replacing https://github.com/aws-cloudformation/cfn-python-lint/blob/cbf972fcb3d45dd71bdac4575e81433cbd36dcc6/src/cfnlint/maintenance.py#L46-L53 with something like 

```python
    # Check to see if we already have the latest version, otherwise download it
    if url_has_newer_version(url):
        spec_content = get_url_content(url, caching=True)
        spec = json.loads(spec_content)
    else:
        with open(filename, 'r') as f:
            spec = json.load(f)
```
This would allow the downloading of updated spec files to be skipped if there are no updates, whilst patching the specs everytime regardless. Alternatively, should we separate the patching of the specs, and include the patching with the scripts you are running?

PatMyron commented:
> Is the problem the fact that `--update-specs` does both the downloading of updated specs, and the merging of the specs with the patches produced by the scripts?

Yeah pretty much. `--update-specs` also merges the specs with manually written patches in addition to patches produced by the scripts

> Alternatively, should we separate the patching of the specs, and include the patching with the scripts you are running?

Need some way for manually written patches to be applied in addition to patches produced by the scripts
